### PR TITLE
added replacement libs

### PR DESCRIPTION
**Pull request by BrandonArp**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+dependency-reduced-pom.xml
+
+# Eclipse
+.metadata/
+**/.classpath
+maven-eclipse.xml
+**/bin/
+**/eclipse-bin/
+**/.externalToolBuilders/
+**/RemoteSystemsTempFiles/
```
Comment (vjkoskela) [inline .gitignore]: Please grab the one from one of the projects that was already updated. These are somewhat draconian versus what is actually needed. For example, suppress all bin directories in the tree!?! I know it's fixed in the parent pom, metrics java client and jvm extra.

File: DEPENDENCIES.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+Project              | License                    | Project link
+---------------------|----------------------------|-------------
+Checkstyle           | LGPL 2                     | http://checkstyle.sourceforge.net/
+FindBugs             | LGPL                       | http://findbugs.sourceforge.net/
+Cobertura            | GPL2                       | http://cobertura.github.io/cobertura/
+Maven                | Apache License 2           | http://maven.apache.org/
+Hamcrest             | BSD 3                      | https://github.com/hamcrest/JavaHamcrest/
+jUnit                | Eclipse Public License 1.0 | https://github.com/junit-team/junit/
+Mockito              | MIT                        | https://code.google.com/p/mockito/
```
Comment (vjkoskela) [inline DEPENDENCIES.md]: Isn't Codehale a dependency of this?

File: DEPENDENCIES.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+
+Compile Dependencies
+--------------------
+
+Project                    | License                    | Project link
+---------------------------|----------------------------|-------------
+Yammer Metrics             | Apache License 2           | http://metrics.codahale.com/
```
Comment (vjkoskela) [inline DEPENDENCIES.md]: Yammer? Really?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,122 @@
...
+Codahale Extra
+============
+
+Extension for clients migrating from Codahale metrics that allows migration to our client library while retaining publication to Codahale.
+
+
+Setup
+-----
+
+### Building
```
Comment (vjkoskela) [inline README.md]: Please move the building section to the bottom. Follow the general structure in the existing extras (e.g. jvm, vertx).

File: README.md [hunk 5]
```diff
@@ -1,2 +1,122 @@
...
+
+```xml
+<dependency>
+    <groupId>com.arpnetworking.metrics.extras</groupId>
+    <artifactId>codahale-extra</artifactId>
+    <version>VERSION</version>
+</dependency>
+```
+
+Add the Maven Central repository either to your ~/.m2/settings.xml or into your project's pom.  Alternatively, if using the local version no changes are necessary as the local repository is enabled by default.
```
Comment (vjkoskela) [inline README.md]: We're removing the local stuff from these notes to gear them more towards end-users. Can you go through this and get it to match? Does the same need to be done to yammer?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,122 @@
...
+
+The Maven Central repository is included by default.  Alternatively, if using the local version add the local repository into project/plugins.sbt:
+
+    resolvers += Resolver.mavenLocal
+
+### Publishing via Codahale
+
+This extra package contains classes that can be used to supplement Codahale output.  The first way this can be done is to use the codahale-extra package and record all 
+metrics against a MetricsRgistry in the com.arpnetworking.metrics.codahale package.  This will allow existing code written against the Codahale metrics interfaces to work 
+ArpNetworking metrics, with the caveat that the use of static methods to create metrics is not supported.  This is the preferred method of using this library.
```
Comment (vjkoskela) [inline README.md]: work with

File: README.md [hunk 7]
```diff
@@ -1,2 +1,122 @@
...
+
+```java
+final MetricsRegistry registry = new com.arpnetworking.metrics.codahale.MetricsFactory();
```
Comment (vjkoskela) [inline README.md]: Shouldn't they inject our MetricsFactory?

File: README.md [hunk 8]
```diff
@@ -1,2 +1,122 @@
...
+
+Codahale metrics provides some interfaces that ArpNetworking metrics does not.  For instance, there is no Meter in ArpNetworking metrics, 
+and counters act differently in Codahale than here.  This section will detail the differences.
+
+##### Counter
+
+Both Codahale and ArpNetworking metrics contain counters that can be incremented by 1 or any arbitrary number.  The difference is in 
+ArpNetworking metrics' separation of units of work and tracking individual samples.  This impedance mismatch is solved by storing each 
+call to Codahale's counter inc as a separate sample in ArpNetworking metrics.  This means that loops where inc() is called multiple times 
+will translate to multiple samples of '1'.  Normally this will not be a problem and the expected value of the Codahale metric will be in
```
Comment (vjkoskela) [inline README.md]: This isn't entirely true. The distributions for the count will be completely off. Would we be better off accumulating all increments and then resetting on publish?

File: codahale-extra/pom.xml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  <artifactId>codahale-extra</artifactId>
+
+  <properties>
+    <!-- Findbugs -->
+    <findbugs.exclude>${project.basedir}/findbugs.exclude.xml</findbugs.exclude>
+  </properties>
+  <build>
+    <resources>
+      <resource>
+        <directory>src/main/resources</directory>
```
Comment (vjkoskela) [inline codahale-extra/pom.xml]: This is unnecessary.

https://maven.apache.org/plugins/maven-resources-plugin/examples/resource-directory.html

File: settings.xml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+      <id>ossrh</id>
+      <username>${env.OSSRH_USER}</username>
+      <password>${env.OSSRH_PASS}</password>
+    </server>
+    <server>
+      <id>arpnet</id>
+      <passphrase>${env.GPG_PASS}</passphrase>
+    </server>
+  </servers>
+</settings>
```
Comment (vjkoskela) [inline settings.xml]: Can you create issues for this and the dw extra to port the reporting class over as well. Doesn't need to be done here.

File: codahale-replace/src/main/java/com/arpnetworking/metrics/Dummy.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics;
+
+/**
+ * Dummy class to trigger javadoc generation.
```
Comment (vjkoskela) [inline codahale-replace/src/main/java/com/arpnetworking/metrics/Dummy.java]: There should be an issue to fix this (and reference here).

File: pom.xml [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+  </developers>
+
+  <modules>
+    <module>codahale-replace</module>
+    <module>codahale-shaded</module>
+    <module>codahale-extra</module>
+  </modules>
+
+  <properties>
+    <main.basedir>${project.basedir}</main.basedir>
```
Comment (vjkoskela) [inline pom.xml]: unused?

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    <maven.test.skip>false</maven.test.skip>
```
Comment (vjkoskela) [inline pom.xml]: unused?

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
```
Comment (vjkoskela) [inline pom.xml]: Both of above are inherited from parent.

File: pom.xml [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    <junit.version>4.12</junit.version>
+    <metrics.client.version>0.3.5</metrics.client.version>
+    <mockito.version>1.10.19</mockito.version>
+    <slf4j.version>1.7.10</slf4j.version>
+
+    <!--Plugin versions-->
+    <maven.shade.plugin.version>2.4.1</maven.shade.plugin.version>
+  </properties>
+  <build>
+    <finalName>${project.artifactId}</finalName>
```
Comment (vjkoskela) [inline pom.xml]: This should not be necessary.

File: DEPENDENCIES.md (hunk 3 above)
Comment (BrandonArp) [inline DEPENDENCIES.md]: Fixed

File: README.md (hunk 5 above)
Comment (BrandonArp) [inline README.md]: Done.

File: README.md (hunk 6 above)
Comment (BrandonArp) [inline README.md]: Fixed

File: README.md (hunk 7 above)
Comment (BrandonArp) [inline README.md]: Fixed

File: codahale-extra/pom.xml (hunk 9 above)
Comment (BrandonArp) [inline codahale-extra/pom.xml]: When I removed this, the LICENSE file did not make it into the jar.

File: settings.xml (hunk 10 above)
Comment (BrandonArp) [inline settings.xml]: Done

File: codahale-replace/src/main/java/com/arpnetworking/metrics/Dummy.java (hunk 11 above)
Comment (BrandonArp) [inline codahale-replace/src/main/java/com/arpnetworking/metrics/Dummy.java]: There is an issue.  Why does every issue need a TODO?

File: codahale-extra/pom.xml (hunk 9 above)
Comment (vjkoskela) [inline codahale-extra/pom.xml]: Are you sure? I just checked the jvm-extra and it doesn't have this and the license file is in the jar. What version of maven are you using? Can you try to upgrade to 3.3.3?

File: .gitignore (hunk 1 above)
Comment (BrandonArp) [inline .gitignore]: Done

File: README.md (hunk 4 above)
Comment (BrandonArp) [inline README.md]: Done

File: README.md (hunk 8 above)
Comment (BrandonArp) [inline README.md]: Done

File: pom.xml (hunk 12 above)
Comment (BrandonArp) [inline pom.xml]: Fixed

File: pom.xml (hunk 13 above)
Comment (BrandonArp) [inline pom.xml]: Fixed

File: pom.xml (hunk 14 above)
Comment (BrandonArp) [inline pom.xml]: Fixed

File: pom.xml (hunk 15 above)
Comment (BrandonArp) [inline pom.xml]: Fixed.

File: codahale-extra/pom.xml (hunk 9 above)
Comment (vjkoskela) [inline codahale-extra/pom.xml]: Discussed and tested; seems to be an artifact of using a multi-module project. Very strange.

**Pull request merged**